sdk_v2/kubeflow/training/api/training_client.py

```diff
@@ -12,7 +12,553 @@
+
+        Raises:
+            ValueError: Input arguments are invalid.
+            TimeoutError: Timeout to create TrainJobs.
+            RuntimeError: Failed to create TrainJobs.
+        """
+
+        if trainer:
+            utils.validate_trainer(trainer)
+
+        # Generate unique name for the TrainJob.
+        # TODO (andreyvelich): Discuss this TrainJob name generation.
+        train_job_name = random.choice(string.ascii_lowercase) + uuid.uuid4().hex[:11]
+
+        # Build the Trainer.
+        trainer_crd = models.KubeflowOrgV2alpha1Trainer()
+
+        # Add number of nodes to the Trainer.
+        if trainer and trainer.num_nodes:
+            trainer_crd.num_nodes = trainer.num_nodes
+
+        # Add resources per node to the Trainer.
+        if trainer and trainer.resources_per_node:
+            trainer_crd.resources_per_node = utils.get_resources_per_node(
+                trainer.resources_per_node
+            )
+
+        # Add command and args to the Trainer if training function is set.
+        if trainer and trainer.func:
+            trainer_crd.command = constants.DEFAULT_COMMAND
+            # TODO: Support train function parameters.
+            trainer_crd.args = utils.get_args_using_train_func(
+                trainer.func,
+                trainer.func_args,
+                trainer.packages_to_install,
+                trainer.pip_index_url,
+            )
+
+        # Add the Lora config to the Trainer envs.
+        if (
+            trainer
+            and trainer.fine_tuning_config
+            and trainer.fine_tuning_config.peft_config
+        ):
+            trainer_crd.env = utils.get_lora_config(
+                trainer.fine_tuning_config.peft_config
+            )
+
+        train_job = models.KubeflowOrgV2alpha1TrainJob(
+            api_version=constants.API_VERSION,
+            kind=constants.TRAINJOB_KIND,
+            metadata=client.V1ObjectMeta(name=train_job_name),
+            spec=models.KubeflowOrgV2alpha1TrainJobSpec(
+                runtime_ref=models.KubeflowOrgV2alpha1RuntimeRef(name=runtime_ref),
+                trainer=(
+                    trainer_crd
+                    if trainer_crd != models.KubeflowOrgV2alpha1Trainer()
+                    else None
+                ),
+                dataset_config=utils.get_dataset_config(dataset_config),
+                model_config=utils.get_model_config(model_config),
+            ),
+        )
+
+        # Create the TrainJob.
+        try:
+            self.custom_api.create_namespaced_custom_object(
+                constants.GROUP,
+                constants.VERSION,
+                self.namespace,
+                constants.TRAINJOB_PLURAL,
+                train_job,
+            )
+        except multiprocessing.TimeoutError:
+            raise TimeoutError(
+                f"Timeout to create {constants.TRAINJOB_KIND}: {self.namespace}/{train_job_name}"
```

Comment:
Doesn't the response have trainjob name?

Reply:
I think, it returns the object which needs to be parsed to get the name. I am not sure if that is necessary here.

Reply:
The primary reason for my question is the pre-defined TrainJob name.
At first glance, if we obtain the name from the response, we do not seem to need to pre-define the TrainJob name, right?

Reply:
I guess, in that case we should use the `generateName` API and pass the default prefix to all TrainJobs.

Additionally, we are discussing with @shravan-achar on how we can use this unique TrainJob ID to snapshot filesystem before the TrainJob is created, so users' files can be places into their distributed ML jobs. But we still need to design the solution for it.

I would like us to come back to this discussion once we have initial support of Kubeflow Training V2 SDK.
E.g. right now, I don't see any invalid generation of the TrainJob's name using this pattern: https://github.com/kubeflow/training-operator/blob/dd3a7a6524b57a6327bd61eec8289204c135b5a8/sdk_v2/kubeflow/training/api/training_client.py#L184
WDYT @tenzen-y ?

Reply:
> Additionally, we are discussing with @shravan-achar on how we can use this unique TrainJob ID to snapshot filesystem before the TrainJob is created, so users' files can be places into their distributed ML jobs. But we still need to design the solution for it.

I see. Can you expand that in the dedicated issue?

Reply:
Sure, created: https://github.com/kubeflow/training-operator/issues/2347